Use tiered projections for hash aggregates

jbrennan333 opened the pull request:
Signed-off-by: Jim Brennan <jimb@nvidia.com>

Closes #5085.

This implements a tiered approach for input projections in hash aagregations.  This is a performance optimization to reduce computation by pulling out common sub-expressions and doing them once.  The general approach is to provide an alternative to bindGpuReferences, currently called bindGpuReferencesTiered, which returns a case class that contains a tiered list of references, and the corresponding method in that case class that applies the tiered projections in order. The tiered list of references are created by recursively pulling out common sub-expressions from the initial list of expressions and rewriting expressions to refer to the output of the sub-expressions.

This feature is enabled by default, but there is a configuration property, `spark.rapids.sql.tiered.project.enabled` that can be used to disable it.

See #5085 for more discussion and performance.  If this works well for aggregations, we will consider enabling it for other projections in the plugin.

jbrennan333 commented:
build

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:
```diff
@@ -0,0 +1,356 @@
+    }
+  }
+
+  def logExprTiers(title: String, tiers: Seq[Seq[Expression]]): Unit = {
```
Are these needed still?

jbrennan333 replied:
I am still using them for debugging, but I will remove before merging.

jbrennan333 replied:
Removed.

revans2 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:
```diff
@@ -0,0 +1,356 @@
+      val equivalentExpressions = new GpuEquivalentExpressions
+
+      // Filter out some expressions that we know are problematic
+      // This won't be necessary if we replace EquivalentExpressions
```
Didn't we replace EquivalentExpressions???

jbrennan333 replied:
Yes. And I removed that filter as well.  I will remove the comment.

jbrennan333 replied:
Comment removed.

abellina commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:
```diff
@@ -0,0 +1,356 @@
+   * Returns the state of the given expression in the `equivalenceMap`. Returns None if there is no
+   * equivalent expressions.
+   */
+  def getExprState(e: Expression): Option[GpuExpressionStats] = {
```
if this is only used in testing, it may be nice to mark it as so, like `getAllExprStates` (nit). Could it also be `private[sql]`?

jbrennan333 replied:
Yes.  I will fix this.

jbrennan333 replied:
done.

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -253,6 +253,41 @@ case class GpuProjectAstExec(
+ * Do projections in a tiered fashion, where earlier tiers contain sub-expressions that are
+ * referenced in later tiers.  Each tier adds columns to the original batch corresponding
+ * to the output of the sub-expressions.
+ */
```
it would be helpful, if there could be minimal example in this header that shows the projected tier process for a couple of tiers and a simple original `cb`. It took me a while to grok how this function works.

jbrennan333 replied:
I will add a comment.

jbrennan333 replied:
Comment added.  I also added a unit test like what the comment says.

abellina replied:
Thanks!

abellina commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/aggregate.scala:
```diff
@@ -708,7 +720,11 @@ class GpuHashAggregateIterator(
     def preProcess(toAggregateBatch: ColumnarBatch): ColumnarBatch = {
       withResource(new NvtxRange("pre-process", NvtxColor.DARK_GREEN)) { _ =>
-        GpuProjectExec.project(toAggregateBatch, preStepBound)
+        if (useTieredProject) {
```
I wish, and it could be follow on work, that this could be handled by `GpuProjectExec.project`. E.g. that `preStepBoundTiered` could be a class, something like `TieredBoundReferences`,  that `.project` could be overloaded for. It would remove some of the changes here, and push the decision to use tiered to the binding part.

jbrennan333 replied:
I agree that would be better.  I may defer this to a follow-on PR though.

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -253,6 +253,41 @@ case class GpuProjectAstExec(
+case class GpuTieredProject(
+    exprSets: Seq[Seq[GpuExpression]]) extends Arm with Logging {
+
+  private def projectTier(boundExprs: Seq[Seq[GpuExpression]], cb: ColumnarBatch): ColumnarBatch = {
```
Yes.  I should be able to make this change.

jbrennan333 replied:
Thanks for the suggestion.  I rewrote it as tail recursive.

jbrennan333 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/catalyst/expressions/GpuEquivalenExpressions.scala:
```diff
@@ -0,0 +1,356 @@
+   * expressions. But when we call `addExprTree` to add common expressions into the map, we
+   * will add recursively the child nodes. So we need to filter the child expressions first.
+   * For example, if `((a + b) + c)` and `(a + b)` are common expressions, we only add
+   * `((a + b) + c)`.
```
This method and comment are from the spark EquivalentExpressions code.  The spark code effectively only gives you one tier of substitutions.  To get multiple tiers, I am calling the spark code recursively in getExprTiers().
In your example, I think the spark code will give you:
```
((a + b) + c), (a + b)
```
I then recurse on this to create another tier, so we end up with:
```
(a+b) as tmp1
(tmp1 + c) as tmp2, tmp1 as tmp3,
tmp2, (tmp2 + d), tmp3
```

revans2 replied:
okay.

jbrennan333 commented on sql-plugin/src/main/scala/com/nvidia/spark/rapids/basicPhysicalOperators.scala:
```diff
@@ -253,6 +253,41 @@ case class GpuProjectAstExec(
+            if (tail.isEmpty) {
+              projectTier(tail, projectCb)
+            } else {
+              withResource(GpuColumnVector.combineColumns(cb, projectCb)) {
```
Agreed.  Removing eclipsed columns would be a good improvement.  It's a little complicated, because the original set of expressions could still have references to a and b that are separate from instances of (a + b), like (a + c) or (a +d).  But we should be able to determine this.   Just adding the subexpression result columns was the easier first step.

revans2 replied:
If you have the tiers before you bind the expressions, then it is not that bad.

Look at the inputs (leaf nodes) of the output tier, this is what the output of the previous tier needs to be. If there is something missing from that tier, then add it in as a separate column that just references that column.  Then go to the next tier and look again.  Repeat until all of the tiers are done.

After that you can bind each tier the regular way to the outputs of the previous tier...

revans2 replied:
We will do this as follow on work so I am fine with putting this in as is now.

jbrennan333 commented on sql-plugin/src/main/scala/org/apache/spark/sql/rapids/AggregateFunctions.scala:
```diff
@@ -802,23 +802,25 @@ case class GpuExtractChunk32(
-          }
-          withResource(zero) { zero =>
-            chunkCol.replaceNulls(zero)
+    withResource(new NvtxRange("GpuExtractChunk32", NvtxColor.RED)) { _ =>
```
I will remove this.

jbrennan333 replied:
Removed.

jbrennan333 commented:
I pushed up a commit that addresses some of the review comments.  I am still doing some additional rework, so please hold off on merging for now.

jbrennan333 commented:
build

jbrennan333 commented:
Updated to address more review comments.  I've done a significant rewrite of getExprTiers to make it more efficient and to fix a problem I was seeing where it was making aliases of aliases in each tier, increasing the number of columns (at least column references) pretty significantly.  It no longer does this.  A simple example is from an earlier comment, these expressions: `((a + b) + c), (a + b)`  produced this set of expression tiers:
```
(a+b) as tmp1
(tmp1 + c) as tmp2, tmp1 as tmp3,
tmp2, (tmp2 + d), tmp3
```
Now this will produce:
```
(a+b) as tmp1
(tmp1 + c) as tmp2
tmp2, (tmp2 + d), tmp1
```
The redundant `tmp1 as tmp3` is removed.  It's small in this example, but in some queries, like q66, there can be hundreds of these.  I don't think it was causing much extra work, because they are all aliases, but it's much cleaner now.

This update does not include removing eclipsed columns.  I will address that in a follow-up PR.

jbrennan333 commented:
build

jbrennan333 commented:
I added some debugging to print out number of columns added and batch sizes for each tier when tiered project is happening.  I ran nds2.0 power run on spark2a and captured this data.  I split it out into three files, one for 2 tiers (1 sub-expression tier), one for 3 tiers (2 sub-expression tiers), and one for 4 tiers (3 sub-expression tiers).  In each of these files, each line looks something like this:
```
cols(in,diff,diff,out)  5  2  2 13      sizes(in,diff,diff,out)           5,044           3,984           7,968          13,510
```
This line shows 5 columns in the input batch, 2 columns added in the first tier, 2 columns added in the second tier, and 13 columns in the output batch.

It also shows a batch size of 5,044 for the input batch and 13,510 for the output batch.
The other batch sizes are differentials from the previous column, so the first tiered project added 3,984 bytes, and the second tier added 7,968 bytes.

The files are sorted by the last differential batch size column.

[batch-data-2-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508334/batch-data-2-tiers-spark2a.txt)
[batch-data-3-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508338/batch-data-3-tiers-spark2a.txt)
[batch-data-4-tiers-spark2a.txt](https://github.com/NVIDIA/spark-rapids/files/9508340/batch-data-4-tiers-spark2a.txt)

jbrennan333 commented:
Thanks for the reviews @revans2 and @abellina!
